Add thread arguments to GuildChannel.edit + applied tags to edit_cannel

ZechCodes opened the pull request:
### Summary
<!-- Small summary of the merge request -->
Implements `GuildThreadChannel.edit` to add support for thread-specific edit fields. Also adds support for applied tags in `RESTClient.edit_channel`.

*MyPy fails because the signature of `GuildThreadChannel.edit` no longer matches the signature of `GuildChannel.edit`. Uncertain how to proceed with that.*

### Checklist
<!-- Make sure to tick all the following boxes by putting an `x` in between (like this `[x]`) -->
- [x] I have run `nox` and all the pipelines have passed.
- [x] I have made unittests according to the code I have added/modified/deleted.

### Related issues
<!--
To mention an issue use `#issue-id` and to mention a merge request use `!merge-request-id`
To close/fix an issue use `Close #issue-id` or `Fix #issue-id` (depending on the merge request)
-->
#1053

ZechCodes commented:
Just to reiterate. I have run Nox.

MyPy fails because the signature of `GuildThreadChannel.edit` no longer matches the signature of `GuildChannel.edit`. Uncertain how to proceed with that.

ZechCodes commented:
#1053 Also this issue is very likely related. I'd be willing to do what's necessary to close it, just need guidance on the next step.

GoogleGenius commented:
> MyPy fails because the signature of `GuildThreadChannel.edit` no longer matches the signature of `GuildChannel.edit`. Uncertain how to proceed with that.

Wouldn't you add these new arguments to the signature for `GuildChannel.edit` instead of overriding in the GuildThreadChannel which violates type safety? Or am I just being very stupid? Please correct me if I'm wrong.

GoogleGenius commented on hikari/channels.py:
```diff
@@ -1696,6 +1696,90 @@ class GuildThreadChannel(TextableGuildChannel):
+        auto_archive_duration : hikari.undefined.UndefinedOr[time.Intervalish]
+            If provided, the new auto archive duration for this thread.
+
+            This should be either 60, 1440, 4320 or 10080 seconds and, as of
```
Actually, I don't know if it's out of the scope of this PR but there are some other docstrings that have the unnecessary "and" I think...

ZechCodes replied:
I see the extraneous "and" is on the doc string for `default_auto_archive_duration` as well. That is likely where I got it from. Removed both.

> The other part of the docstring that this was copied from might be important as well...

Not sure what you are referring to.

GoogleGenius replied:
Ignore that comment, sorry!

davfsa commented on changes/1474.bugfix.md:
```diff
@@ -0,0 +1,2 @@
+- Implement `GuildThreadChannel.edit`
+- Update `RESTClient.edit_channnel` to support applied tags on forum threads
```
This one would not fall as a bugfix, but as a feature rather

ZechCodes replied:
👍

davfsa commented on hikari/channels.py:
```diff
@@ -1696,6 +1696,90 @@ class GuildThreadChannel(TextableGuildChannel):
     Will be `None` for threads created before 2020-01-09.
     """
 
+    async def edit(
+        self,
+        *,
+        name: undefined.UndefinedOr[str] = undefined.UNDEFINED,
+        archived: undefined.UndefinedOr[bool] = undefined.UNDEFINED,
+        auto_archive_duration: undefined.UndefinedOr[time.Intervalish] = undefined.UNDEFINED,
+        locked: undefined.UndefinedOr[bool] = undefined.UNDEFINED,
+        invitable: undefined.UndefinedOr[bool] = undefined.UNDEFINED,
+        rate_limit_per_user: undefined.UndefinedOr[time.Intervalish] = undefined.UNDEFINED,
+        flags: undefined.UndefinedOr[ChannelFlag] = undefined.UNDEFINED,
+        applied_tags: undefined.UndefinedOr[typing.Sequence[ForumTag]] = undefined.UNDEFINED,
+        reason: undefined.UndefinedOr[str] = undefined.UNDEFINED,
+    ) -> PartialChannel:
```
We cant abide by the issue you linked this PR to yet, as it will (as you found out) create conflicts on mypy. The best to do here is to just temporarily add the thread fields to `GuildChannel.edit`. You can also re-scope this PR and complete that issue if you wish, by properly splitting the arguments in the different objects, which would fix the mypy issue

ZechCodes replied:
That's why I left things as they are in the PR. I wasn't sure if someone might have a better idea of how to approach it.

One thought I had is that threads, dms, and channels are all distinct types and only share some behavior. So one idea I had was maybe to create a base channel type for the shared behavior, then branch that to dm channel, guild channel, and thread subtypes that implement the respective behaviors. That's a more extensive change though which is beyond the scope of what I should be doing without input. Will almost certainly be a breaking change for someone.

So, I'm gonna merge the thread behavior into the `GuildChannel.edit` and remove the `GuildThreadChannel.edit`. I would be open to helping break up the channel types, though, if there were interest in going through that process.

davfsa replied:
>  I would be open to helping break up the channel types, though, if there were interest in going through that process.

Yeah, that would be nice :)

davfsa commented on tests/hikari/test_channels.py:
```diff
@@ -383,6 +383,61 @@ async def test_edit(self, model):
+        )
+
+    @pytest.mark.asyncio()
+    async def test_thread_edit(self, model):
```
```suggestion
    async def test_edit(self, model):
```

ZechCodes replied:
I've removed that test as it is no longer necessary, given I'm merging `GuildThreadChannel.edit` into `GuildChannel.edit`.

davfsa commented on hikari/impl/rest.py:
```diff
@@ -1054,6 +1055,7 @@ async def edit_channel(
         body.put("auto_archive_duration", auto_archive_duration, conversion=time.timespan_to_int)
         body.put("locked", locked)
         body.put("invitable", invitable)
+        body.put_array("applied_tags", applied_tags, conversion=self._entity_factory.serialize_forum_tag)
```
This is missing a test case

ZechCodes replied:
That should be covered by the `TestGuildThreadChannel.test_thread_edit` test. However, as I am removing the `GuildThreadChannel.edit` method, I will also remove that test. In its place, I will make the necessary changes to the `TestGuildChannel.test_edit` test.

ZechCodes replied:
Realized there are tests for impl. Updating the tests.

davfsa commented on hikari/impl/rest.py:
```diff
@@ -1003,6 +1003,7 @@ async def edit_channel(
         locked: undefined.UndefinedOr[bool] = undefined.UNDEFINED,
         invitable: undefined.UndefinedOr[bool] = undefined.UNDEFINED,
         auto_archive_duration: undefined.UndefinedOr[time.Intervalish] = undefined.UNDEFINED,
+        applied_tags: undefined.UndefinedOr[typing.Sequence[channels_.ForumTag]] = undefined.UNDEFINED,
```
Ty for spotting that this is missing!

ZechCodes replied:
Np

ZechCodes commented on hikari/channels.py:
```diff
@@ -1696,6 +1696,90 @@ class GuildThreadChannel(TextableGuildChannel):
+            If provided, the new channel flags to use for the thread. PINNED can
+            only be used on threads in forum channels.
+        applied_tags : hikari.undefined.UndefinedOr[typing.Sequence[ForumTag]]
+            If provided, the new tags to use for the thread in a forum channel.
```
👍

ZechCodes commented:
I've merged what I had in `GuildThreadChannel.edit` into `GuildChannel.edit`. I've removed the test I added for `GuildThreadChannel.edit` and updated the existing test for `GuildChannel.edit` to cover the new additions.

I've also updated the test for `RESTClient.edit_channel` to cover both the `flags` and the new `applied_tags` parameters.

GoogleGenius commented on hikari/api/rest.py:
```diff
@@ -306,6 +307,10 @@ async def edit_channel(
             writing, ignores the parent channel's set default_auto_archive_duration
             when passed as `hikari.undefined.UNDEFINED`.
+        applied_tags : hikari.undefined.UndefinedOr[typing.Sequence[hikari.channels.ForumTag]]
+            If provided, the new tags applied to the thread.
```
Slight nitpick.
```suggestion
            If provided, the new tags to apply to the thread.
```
Also, it might be worth removing the spacing between the following statement for consistency with the docstring in the edit method of GuildChannel.

ZechCodes replied:
Are you referring to the spacing in this section?
```
If provided, the new auto archive duration for this thread. This
only applies to threads.

This should be either 60, 1440, 4320 or 10080 seconds, as of
writing.
```

GoogleGenius replied:
Lines 310-313:
```
applied_tags : hikari.undefined.UndefinedOr[typing.Sequence[hikari.channels.ForumTag]]
    If provided, the new tags applied to the thread.

    This only applies to threads in a forum channel.
```
Becomes:
```
applied_tags : hikari.undefined.UndefinedOr[typing.Sequence[hikari.channels.ForumTag]]
    If provided, the new tags applied to the thread. This only applies
    to threads in a forum channel.
```

ZechCodes replied:
Ah, yep, I see. I went ahead and made all the new docs more consistent between `RESTClient.edit_channel` and `GuildChannel.edit`.

davfsa commented on hikari/api/rest.py:
```diff
@@ -242,7 +243,9 @@ async def edit_channel(
             If provided, the new name for the channel.
         flags : hikari.undefined.UndefinedOr[hikari.channels.ChannelFlag]
-            If provided, the new flags for the channel
+            If provided, the new channel flags to use for the channel. This can
+            only be used on a forum channel to apply ChannelFlag.REQUIRE_TAG, or
+            on a forum thread to apply ChannelFlag.PINNED.
```
```suggestion
            If provided, the new channel flags to use for the channel. This can
            only be used on a forum channel to apply `REQUIRE_TAG`, or
            on a forum thread to apply `PINNED`.
```

davfsa commented on hikari/api/rest.py:
```diff
@@ -291,21 +294,25 @@ async def edit_channel(
-
-            This only applies to private threads.
+            If provided, the new setting for whether non-moderators can invite
+            new members to a private thread. This only applies to threads.
```
Why was the wording changed for this one?

ZechCodes replied:
Make the doc consistent with the docs for `GuildChannel.edit` which is taken directly from the Discord Resources docs.

davfsa replied:
I think i like the other wording better, but would appreciate other comments :)

davfsa reviewed (commented):
Ty for working on this and spotting the missing arguments!
